Add timeout parameter to win_service functions

twangboy opened the pull request:
### What does this PR do?
Adds a timeout parameter to the following functions in win_service:

- start
- stop
- restart
- delete

Fix some documentation formatting issues

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/48163

### Tests written?
No

### Commits signed with GPG?
Yes

twangboy committed:
Add timeout parameter

Add a timeout parameter to start, stop, restart, and delete

twangboy committed:
Add timeout support to the state

Remove redundent net stop command from the stop function
Remove extra _enable call on Windows that was causing incorrect state
return comments

damon-atkins commented on salt/modules/win_service.py:
```diff
@@ -409,38 +411,33 @@ def start(name):
     attempts = 0
     while info(name)['Status'] in ['Start Pending', 'Stopped'] \
-            and attempts <= RETRY_ATTEMPTS:
+            and attempts <= timeout:
```
The time out is doco as seconds not attempts please use the following code 

```
    end_t = time.time() + timeout 
    while info(name)['Status'] in ['Start Pending', 'Stopped'] \
          and time.time() < end_t:
        time.sleep(1)
```

damon-atkins commented on salt/modules/win_service.py:
```diff
@@ -450,26 +447,35 @@ def stop(name):
     attempts = 0
     while info(name)['Status'] in ['Running', 'Stop Pending'] \
-            and attempts <= RETRY_ATTEMPTS:
+            and attempts <= timeout:
```
dito

damon-atkins commented on salt/modules/win_service.py:
```diff
@@ -1282,7 +1319,7 @@ def delete(name):
 
     attempts = 0
-    while name in get_all() and attempts <= RETRY_ATTEMPTS:
+    while name in get_all() and attempts <= timeout:
```
dito

damon-atkins reviewed (changes requested):
Please use  "time code" for timeout rather than attempts

damon-atkins committed:
2017_win_service_damon

Merge this if you like the suggestion.

damon-atkins commented:
open a PR against your fix.

twangboy committed:
Merge pull request #10 from damon-atkins/patch-1

2017_win_service

twangboy committed:
Fix start/stop functions

Fix typo in the dict key 'Status'
Standardize docs
Fix some lint issues
Rename the new helper function to _status_wait
Add urls to code comments
Add version added statements

twangboy committed:
Remove testing debug stuff

twangboy committed:
Improve timeout in delete

Improved error handling

twangboy committed:
Fix some lint (remove whitespace)`

damon-atkins commented on salt/modules/win_service.py:
```diff
@@ -100,6 +98,50 @@ def __virtual__():
+        # than 10 seconds.
+        # https://docs.microsoft.com/en-us/windows/desktop/services/starting-a-service
+        # https://docs.microsoft.com/en-us/windows/desktop/services/stopping-a-service
+        wait_time = info_results['Status_WaitHint'] / 1000
```
Looks like I got this code slightly wrong.
Should not have the / 1000   see the else which should do this. This way we cover the off chance wait hint is 0.

``
else:
           wait_time = wait_time / 10000  # convert ms to seconds, and then divide by 10.
``

twangboy committed:
Work with seconds

twangboy commented:
@damon-atkins Could you please review this one last time. I implemented the changes you suggested.

damon-atkins reviewed (approved):
DONE PR update with some more words around the timeout
